brml/swap/src/lib.rs

```diff
@@ -1,1022 +1,1400 @@
+	/// tAo = tokenAmountOut         /    /                            \               \
+	/// bO = tokenBalanceOut        /    /  /           tAo            \\              \
+	/// pAo = poolAmountOut pAo =  | 1 -|1-| -------------------------- || ^  (wO / tW) | * ps
+	/// ps = poolSupply             \    \  \ bO * (1-(1- wO/tW) * sF) //              /
+	/// wO = tokenWeightOut          \    \                            /              /
+	/// tW = totalWeight
+	/// sF = swapFee
+	///
+	///**************************************************************************************************/
+	pub(crate) fn calculate_pool_in_given_single_out(
+		token_weight_out: T::PoolWeight,
+		token_amount_out: T::Balance,
+		token_balance_out: T::Balance,
+		pool_supply: T::Balance,
+		swap_fee: T::Fee,
+	) -> Result<I64F64, Error<T>> {
+		// type convert to u128
+		let token_amount_out =
+			TryInto::<u128>::try_into(token_amount_out).map_err(|_| Error::<T>::ConvertFailure)?;
+		let token_balance_out =
+			TryInto::<u128>::try_into(token_balance_out).map_err(|_| Error::<T>::ConvertFailure)?;
+		let pool_supply =
+			TryInto::<u128>::try_into(pool_supply).map_err(|_| Error::<T>::ConvertFailure)?;
+		let swap_fee =
+			TryInto::<u128>::try_into(swap_fee).map_err(|_| Error::<T>::ConvertFailure)?;
+
+		// to fixed num
+		let token_balance_out = I64F64::from_num(token_balance_out);
+		let token_amount_out = FixedI128::<extra::U64>::from_num(token_amount_out);
+		let pool_supply = FixedI128::<extra::U64>::from_num(pool_supply);
+		let swap_fee = {
+			let precision = TryInto::<u128>::try_into(T::FeePrecision::get())
+				.map_err(|_| Error::<T>::ConvertFailure)?;
+			let precision = FixedI128::<extra::U64>::from_num(precision);
+			FixedI128::<extra::U64>::from_num(swap_fee).saturating_div(precision)
+		};
+
+		let weight_ratio = Self::weight_ratio(token_weight_out, T::WeightPrecision::get())?;
+
+		let inside_part_rhs_denominator_multiplier = FixedI128::<extra::U64>::from_num(1)
+			.saturating_sub(FixedI128::<extra::U64>::from_num(1).saturating_mul(swap_fee));
+		let inside_part_rhs = token_amount_out
+			.saturating_div(token_balance_out)
+			.saturating_div(inside_part_rhs_denominator_multiplier);
+		let inside_part = FixedI128::<extra::U64>::from_num(1).saturating_sub(inside_part_rhs);
+		let power: FixedI128<extra::U64> = transcendental::pow(inside_part, weight_ratio)
+			.map_err(|_| Error::<T>::FixedPointError)?;
+
+		let pool_amount_out =
+			pool_supply.saturating_mul(FixedI128::<extra::U64>::from_num(1).saturating_sub(power));
+		Ok(pool_amount_out)
+	}
+
+	// ********************************************************************************
+	// below are the interfaces needed from other pallets.
+	// Query for the current bonus balance for the pool
+	pub(crate) fn get_bonus_pool_balance(_pool_id: T::PoolId) -> T::Balance {
+		T::Balance::from(1_000_000) // to get from other pallets. Not yet implemented
+	}
+}
+
+#[allow(dead_code)]

```

Comment:
Removing line 1385 to 1400 would be better.
It's already deprecated.

Reply:
fixed.